Markdown html process links images

fiskus opened the pull request:
* Added 'processed' data attribute to links and images, processed with markdown
* Process the rest of the images and links (written using HTML) with the same functions

nl0 commented on catalog/app/components/Markdown/Markdown.tsx:
```diff
@@ -189,7 +189,7 @@ const linkHandler =
       const rel = nofollow ? ' rel="nofollow"' : ''
       return `<a href="${(Remarkable as unknown as RemarkableWithUtils).utils.escapeHtml(
```
let's adjust the formatting here a little bit: move `(Remarkable as unknown as RemarkableWithUtils).utils.escapeHtml(...` to a separate line and assign it to a variable for readability (i know it's kinda out of scope, but anyways, let's "fix" it since we're touching it)

nl0 commented on catalog/app/components/Markdown/Markdown.tsx:
```diff
@@ -219,6 +246,8 @@ export const getRenderer = memoize(({ images, processImg, processLink }) => {
   md.use(imageHandler({ disable: !images, process: processImg }))
```
since we're handling all images and links in the purify's hooks, we can (and should) remove this logic from markdown handlers so we're not doing it twice

fiskus replied:
This makes sense, but I have some resistance, because first we do "right" handling using some AST, and then "hacky" handling using HTML elements mutation

nl0 replied:
well, yeah, but
1. we do that "hacky" thing anyways
2. the "right" way doesn't handle html insertions, which are a supported part of the markdown standard, so it's incomplete

also, remarkable's AST thing looks like a joke TBH with all these string manipulations and lists of tokens...

nl0 reviewed (approved):
looks good, and the solution (using purify's hooks) is very smart, i haven't thought about that. some minor suggestions inline

fiskus commented:
* Replaced processing of the markdown by `dompurify` hook, and made it in one loop cycle
* Dropped most of the escaping, because `setAttribute` is safe enough
* Simplified data structure of the link/image processors: `({ href: string }) => { href: string }` --> `(href: string) => string`
* Removed unused functions and component arguments

nl0 commented on catalog/app/components/Markdown/Markdown.tsx:
```diff
@@ -113,114 +112,86 @@ const highlight = (str: string, lang: string) => {
+  }
+  const title = element.getAttribute('title')
+  if (title) {
+    element.setAttribute('title', title)
```
what's the purpose of getting the attribute and then setting it to the same value?

fiskus replied:
It was leftover, when there was an escaping
I suppose, now, we don't need an escaping because we don't use strings for creating HTML elements

nl0 commented on catalog/app/components/Markdown/Markdown.tsx:
```diff
@@ -113,114 +112,86 @@ const highlight = (str: string, lang: string) => {
+
+function handleLink(process: AttributeProcessor, element: HTMLElement): Element {
+  const attributeValue = element.getAttribute('href')
+  if (!attributeValue) return element
```
if we short-circuit here, `setAttribute` calls below won't be executed. is this intentional?

fiskus replied:
Yes, it was. But I found that we process `href=""` (as relative path to the root of the package), so I changed this condition to `if (typeof attributeValue !== 'string')`

fiskus commented on catalog/app/components/Markdown/Markdown.tsx:
```diff
@@ -272,21 +243,19 @@ export function Container({ className, children }: ContainerProps) {
 
 export default function Markdown({
   data,
-  images = true,
```
Yes, it's never set to `false`, only `true` or omitted

fiskus replied:
It was `false` for
a/catalog/app/containers/Package/Comment.js
b/catalog/app/containers/Package/CommentForm.js

nl0 commented:
it would be nice to add some tests if you have enough capacity and enthusiasm

fiskus commented:
I'm going to add unit-test for the `getRenderer` in the follow-up PR with slightly changed API for `Markdown#getRenderer`

What is really worth is unit-test for `Markdown` component, but, yeah, this is above my capacity :)
